Implement serialization/deserialization

ejpbruel2 opened the pull request.

ejpbruel2 committed:
Implement serialization/deserialization

jochemborges commented on vm/src/deserializer.h:
```diff
@@ -0,0 +1,83 @@
+
+namespace Borges {
+
+class Deserializer {
```
I'd like some basic comments in the code before review, like "deserialization class, copies double* (bytecode?) array and constructs..." - this way you send me into the woods without light (or clothes).

ejpbruel2 replied:
It's important to note here that the Deserializer does *not* copy the buffer. Rather, it stores a reference to the buffer. The same is true for Serializer. The idea is that you can wrap a buffer in a serializer, use the serializer to write some values to it, and then hand the buffer off to somewhere else, like so:

std::vector<double> b;
Serializer s(b);
serialize(s, true);
serialize(s, 42.0);
return b;

jochemborges replied:
yes of course, that makes sense. but then, this would be an awesome comment block :)

ejpbruel2 replied:
Agreed. I'll add a comment here.

jochemborges commented on vm/src/deserializer.h:
```diff
@@ -0,0 +1,83 @@
+private:
+    std::size_t size_;
+    const double* data_;
+    std::size_t index_;
```
what is 'index_' ? 

basic inline comments are welcome to avoid unnecessary review back-and-forth

ejpbruel2 replied:
Index is the current index into the buffer.

jochemborges replied:
and the index is implicitly incremented I see.
would be a good comment in the code :)

ok if added to the code as comment

ejpbruel2 replied:
Will add a comment for this.

ejpbruel2 replied:
Instead of writing a comment here, I opted to rename index_ to bufferIndex_.

The real problem here was that index_ is too generic a name, so it was unclear from the context what it was an index *for*. Giving it a more meaningful name like bufferIndex_ should go a long way in making the code self-documenting (which, in my opinion, should be preferred over comments when possible).

jochemborges commented on vm/src/deserializer.h:
```diff
@@ -0,0 +1,83 @@
+    std::unordered_map<std::size_t, std::shared_ptr<void>> index_to_ptr_map_;
+};
+
+void deserialize(Deserializer& d, bool* p);
```
why do you create global functions here ? for convenience ?

ejpbruel2 replied:
As an overall design pattern, I'm trying to establish an invariant for each class with its constructor, and then try to maintain that variant by keeping the API surface for that class as small as possible. That is, by keeping the number of member functions as small as possible, and defining all other functionality in terms of these member functions, the number of functions that can potentially break these invariants is reduced, which significantly aids debugging.

jochemborges replied:
ok

jochemborges commented on vm/src/deserializer.h:
```diff
@@ -0,0 +1,83 @@
+
+template<typename T>
+void
+deserialize(Deserializer& d, std::unique_ptr<T>* p)
```
why don't you pass by (const) reference here ?

ejpbruel2 replied:
Because deserialize functions always call methods that need to change the state of the deserializer (most of the time, bumping the index into the buffer), and thus are non-const.

jochemborges replied:
no, i mean `std::unique_ptr<T>* p` instead of `const std::unique_ptr<T> & p` - probably because of the new (p) magic seen before.

ejpbruel2 replied:
Exactly. Placement new requires a pointer. It can't use a reference. In general, I almost always use references, except when dealing with low level memory apis like this.

jochemborges replied:
yup, valid reason, thx.

jochemborges commented on vm/src/deserializer.cpp:
```diff
@@ -0,0 +1,42 @@
+deserialize(Deserializer& d, bool* p)
+{
+    *p = deserialize<double>(d);
+}
```
shouldn't this be `deserialize<bool>(d)` ?

ejpbruel2 replied:
No, the template argument determines what type of value we should attempt to read from the buffer. Booleans are serialized as doubles, so they should be deserialized as doubles. The assignment to *p implicitly casts the double to bool.

jochemborges replied:
ah. in that case, an explicit cast is more clear.

ejpbruel2 replied:
I disagree. It would also be redundant, since you'd still have to write:

    static_cast<double>(deserialize<double>(d))

jochemborges replied:
you'd have to write `static_cast<bool>deserialize<double>(d)`, right? or `(bool)deserialize<double>(d)`.

jochemborges commented on vm/src/deserializer.cpp:
```diff
@@ -0,0 +1,42 @@
+void
+deserialize(Deserializer& d, std::size_t* p)
+{
+    *p = deserialize<double>(d);
```
shouldn't this be `deserialize<std::size_t>(d)` ?

ejpbruel2 replied:
Similarly as above: std::size_t is serialized as a double, so should be deserialized as a double, then cast to std::size_t.

jochemborges replied:
similar comment as above,  an explicit cast is preferred

ejpbruel2 replied:
See response above.

jochemborges replied:
ok, i yield to your style

jochemborges commented on vm/src/deserializer.h:
```diff
@@ -0,0 +1,83 @@
+class Deserializer {
+public:
+    Deserializer(std::size_t n, const double* p);
+    void deserialize(double* p);
```
why is the deserialize with (double *p) a separate function (and not implemented via the shared_ptr variant ?)

ejpbruel2 replied:
This is a good question, and I can see why this would be non obvious. To answer this question, I need to establish some context first, so please bear with me.

Suppose we want to be able to deserialize values like this:

    std::vector<double> xs = deserialize<std::vector<double>>(d)

To achieve this, we could define deserialize as a template function, and then specialize this template for each deserializable type. For instance, we can write:

    template <typename T> T deserialize(Deserializer& d);

If we then want to specialise that function for different types, we would write:

    template <> bool deserialize<bool>(Deserializer& d);
    template <> bool deserialize<double>(Deserializer& d);

So far, so good.

Unfortunately, the above approach breaks down when we want to deserialize more complex types. The first problem is that C++ does not allow partial specialization of template functions. That is, we cannot specialize deserialize for, say, vectors of an arbitrary type:

    template <typename T>
    std::vector<T> deserialize<std:vector<T>>(Deserializer& d); // THIS IS NOT ALLOWED!

The second problem is that to deserialize a value of type T, the deserialize function needs to create a temporary variable tmp of type T, into which the value will be read. This is not a problem for POD-types like bool or double, but doesn't work for types that don't have a default constructor, such as Curve. In that case, writing T tmp; will lead to a compiler error.

We could require that each serializable type needs to have a default constructor, but that's an artificial restriction, and still doesn't solve the problem of partial template function specialization. Instead, we will tackle both the above problems at the same time, using the following approach:

Instead of creating a temporary value of type T, deserialize creates a buffer of uninitialized memory thats large enough to contain a value of type T (this buffer needs to be properly aligned, so we use the alignof operator to achieve this). We then use placement new to directly construct a value of type T in this uninitialized memory, using the appropriate constructor. By directly constructing the value in uninitialized memory instead of assigning it to a temporary value, we avoid a needless call to the (possibly non-existent) default constructor.

To solve the problem of partial template function specialization, we introduce a level of indirection: instead of initialising the above pointer directly, deserialize calls a second deserialize function that takes the uninitialized pointer as an argument. This second function is responsible for doing the actual deserialization and constructing the value. Crucially, because this second function is *not* a template, it can be freely overloaded for any type T we want.

Since overloading a function with a template is not the same thing as specializing a template function, we can now define a function that deserializes, say, a vector of an arbitrary type, like this:

    template <typename T>
    std::vector<T> deserialize(Deserializer& d, std::vector<double>* p); // THIS IS ALLOWED!

With the above definition in place, the following would work, which is what we originally set out to do:

    std::vector<double> xs = deserialize<std::vector<T>>(d)

ejpbruel2 replied:
Now that I've explained the context of what we're trying to do, let me get back to your original question. What you are looking at is a member function that deserializes and constructs a double, using the given pointer. In contrast, the shared_ptr variant deserializes and constructs a shared_ptr, using the given pointer.

As an aside, these are the only two functions that need direct access to the state of the Deserializer, which is why they are member functions. All the other functions can be implemented in terms of deserializing doubles or shared_ptrs.

jochemborges replied:
ok

jochemborges commented on vm/src/deserializer.h:
```diff
@@ -0,0 +1,83 @@
+deserialize(Deserializer& d)
+{
+    alignas(T) char b[sizeof(T)];
+    T* p = reinterpret_cast<T*>(b);
```
I would write this as reinterpret_cast<T*>(&b[0]) for clarity

ejpbruel2 replied:
That seems to suggest that p is a pointer to the first element in the buffer, whereas it should more properly be thought of as a pointer to the entire buffer. Would this work for you as an alternative?

    reinterpret_cast<T*>(&b)

jochemborges replied:
no, because b is already a char *. you actually *do* want a pointer to the first element. this way, you'd take the address of the address of b.

ejpbruel2 replied:
You are right, but in that case I disagree that reinterpret_cast<T*>(&b[0]) adds clarity. In fact, it does the opposite for me: you correctly pointed out that b is already a char*, and this simply a pointer cast, but its precisely the change you proposed that got me confused about this fact.

jochemborges replied:
to me, it *is* more clear because you make explicit that yo take the address of the beginning of the array. just passing b is basically an implicit cast to * of the value of b on the stack which is not immediately clear. but again, this is mostly style discussion and I tend to yield to your style (as you must code it ;)

jochemborges commented on vm/src/deserializer.h:
```diff
@@ -0,0 +1,83 @@
+    T* p = reinterpret_cast<T*>(b);
+    deserialize(d, p);
+    return *p;
+}
```
aren't you returning a value on the stack here ?

ejpbruel2 replied:
This is a bit tricky.

As I explained earlier, we create a buffer b to uninitialized storage here, and then take a pointer p to that storage. The call to deserialize on line 35 deserializes and constructs a value in the uninitialized storage pointed to by p. In other words, after this call, p now points to a value of type T, which is stored in the buffer b. When we return *p, what we actually return is a *copy* of the value pointed to by p.

However, you did make me realise a subtle bug in this code. When we return from this function, the buffer b goes out of scope, but we never call the destructor for the value that it stores. That's a potential memory leak if the value in question owns any pointers that it needs to release in it's destructor.

To avoid this, we should first copy the value stored in the buffer b to a temporary value v, explicitly call the destructor for the value in the buffer, and *then* return a copy of v (unlike the value stored in but, the destructor for v will be called automatically when we go out of scope).

As an extra optimization, we should use move instead of copy semantics here to avoid potentially expensive copying. I will update the pull request accordingly.

jochemborges commented on vm/src/deserializer.h:
```diff
@@ -0,0 +1,83 @@
+    }
+    std::shared_ptr<T> ptr(std::make_shared<T>(deserialize<T>(*this)));
+    index_to_ptr_map_.insert(std::make_pair(i, std::shared_ptr<void>(ptr)));
+    new (p) std::shared_ptr<T>(std::move(ptr));
```
thislast line (51) is magic to me, please explain (in comments)

ejpbruel2 replied:
Line 51 constructs a value in the uninitialized storage pointed to by p, using placement new (this is a C++ feature, see http://en.cppreference.com/w/cpp/language/new).

The other lines are simpler. Since this is the first time we've seen this shared_ptr, on line 49 we deserialize the value that it points to, and create a shared_ptr from that (the use of make_shared is an optimization that allocates the refcount and the value itself in a single block). Then on line 50, we associate the index we deserialized earlier with the shared_ptr we just created. If we encounter the same index when deserializing a shared_ptr again, we return the shared_ptr associated with that index instead of deserializing the value again (this is what allows us to preserve shared_ptrs across serialization!).

jochemborges replied:
sorry, i made a typo, or the lines have moved, i meant the line

`new (p) std::shared_ptr<T>(std::move(ptr));`

ejpbruel2 replied:
That's placement new syntax, which is part of C++03. It works exactly like the new operator you are used to, except that it doesn't allocate memory for the object to be constructed. Instead, it uses the memory pointed to by p to construct the shared_ptr in.

jochemborges replied:
ok, i see.

ejpbruel2 commented on vm/src/deserializer.cpp:
```diff
@@ -0,0 +1,42 @@
+}
+
+void
+Deserializer::deserialize(double* p)
```
Test

jochemborges commented on vm/src/deserializer.cpp:
```diff
@@ -0,0 +1,42 @@
+}
+
+void
+deserialize(Deserializer& d, double* p)
```
why does a void function use the return statement here ?

ejpbruel2 replied:
I think that's a leftover from an earlier iteration when Deserializer:deserialize still returned something. I agree that it should be removed.

jochemborges commented on vm/src/serializer.h:
```diff
@@ -0,0 +1,68 @@
+    serialize(i);
+    serialize(*p);
+    ptr_to_index_map_.insert(std::make_pair(std::shared_ptr<void>(p), i));
+}
```
why do you serialize the size of the index map ?

ejpbruel2 replied:
The fact that I used i and not n as the name for the variable should have been a hint that this is an index, not a size. I use the size of the map for the value of the index because I need a unique index that has not been used before. If that isn't clear from this code, I will add a comment to this effect.

jochemborges replied:
i read it as an index, but noticed the value, which made me think. I usually use -1 to indicate that the value has not been used or 0(zero) if all valid indices are > 0. otherwise, indeed, please comment.

jochemborges commented on vm/src/deserializer.h:
```diff
@@ -0,0 +1,85 @@
+    void deserialize(std::shared_ptr<T>* p);
+
+private:
+    std::size_t size_;
```
i much prefer prefix notation by the way, eg. m_size for member variables, g_xxx for globals ans s_static for class variables. this way, the trailing_ is easily overlooked.

ejpbruel2 replied:
I've adopted this style from Mozilla's C++ codebase, and have not ever found it to be a problem in practice. I generally avoid using prefixes like g_ and s_ because I don't think Hungarian notation (i.e. encoding type information in the variable name) is a good idea. It makes things harder to refactor should I ever decide to change the scope of a variable later on.

jochemborges replied:
the refactoring is a valid argument, but what i propose is that you show only the scope by using m_, g_ and s_ prefixes, not the type. refactoring issues are therefore irrelevant. Anyway, being comfortable in your coding style is important too, so once again, I yield ;)

jochemborges replied:
i wasn't proposing "full" hungarian notation, just adding the scope (m_,g_ and s_) so refactoring is not an issue. for me, it reads much easier, as it's a prefix, not a suffix. anyway, if you find it too much effort stick to the proven routine.

jochemborges commented on vm/src/serializer.cpp:
```diff
@@ -0,0 +1,33 @@
+{
+    s.serialize(x);
+}
+
```
still, after our discussions, I don't understand why these cannot or shouldn't be member functions. each function manipulates the state of the Serializer, right? then why whould these be global functions? to align with the deserialize API?

ejpbruel2 replied:
No. None of these functions manipulate the state of the serializer directly. Only the two member functions (for serializing doubles or shared_ptrs) can do that. They are thus the only ones that can possible mess up the invariants for the class. All the other functions that use the serializer are implemented *in terms* of these member functions, so they don't need direct access (and hence should not be members);

ejpbruel2 commented:
Okay, I *think* I've addressed all your review comments now. In particular, I've added several comments that should make the things you were confused about a lot clearer. Let me know if this suffices for an r+

jochemborges commented:
r+ with comments addressed

ejpbruel2 commented:
@jochemborges Which comments still need addressing? I'm unsure.

ejpbruel2 commented:
Just confirmed with @jochemborges on Slack that this pr can be merged.
